### PR TITLE
[WIP] visitor pattern for the ast

### PR DESCRIPTION
**Pull request by L3nn0x**
Implements a simple read-only visitor to answer #4

### CODE REVIEW

**Comment by L3nn0x** [comment]
I'm not sure about the default order of visit.

**Comment by vickenty** [comment]
Thank you, looks good.

As for visiting order, In general, I think the visiting order should match the order elements appear in the C source. Most of the structs look fine in this regard, except maybe `DoWhileStatement`, which should visit the statement first.

It looks that most parameter names in the module start with an underscore even though they are used. Is it on purpose?

Do you think default branches are necessary when matching enum's?  it might be easier to keep visitor in sync with ast if we leave them out.

Finally, did you use any tool to generate it, or did you write it by hand? If there is a tool, perhaps it can be integrated into the Makefile.

**Comment by L3nn0x** [comment]
I changed the visit order of `DoWhile` and `UnaryOperator` to match the order of the C code.

File: src/visitor.rs [hunk 1]
```diff
@@ -2,612 +2,620 @@ use ast::*;
...
             for c in comma.iter() {
-                _visitor.visit_expression(&c.node, &c.span);
+                visitor.visit_expression(&c.node, &c.span);
             }
         },
-        Expression::OffsetOf(o) => _visitor.visit_offset_of_expression(&o.node, &o.span),
-        Expression::VaArg(v) => _visitor.visit_va_arg_expression(&v.node, &v.span),
-        Expression::Statement(s) => _visitor.visit_statement(&s.node, &s.span)
+        Expression::OffsetOf(o) => visitor.visit_offset_of_expression(&o.node, &o.span),
+        Expression::VaArg(v) => visitor.visitva_arg_expression(&v.node, &v.span),
```
Comment (vickenty) [inline src/visitor.rs]: Looks like an extra underscore got removed between `visit` and `va`.

**Comment by vickenty** [comment]
The code in general looks good. It would be perfect if you could add just a few words for the module-level documentation, but I can do it later myself too. I'm going to merge this as is once the `visitva` typo is fixed.

Thanks for implementing this, good job!

**Comment by L3nn0x** [comment]
I've never been good at adding documentation, but I can give it a try.

**Comment by vickenty** [comment]
The code looks good now, so I've merged it as is. Thanks for the contribution!

Docs still would be nice, so if you feel like adding some, feel free to open new PR. Otherwise I'd probably get around to it myself next weekend.

**Pull request merged**